### PR TITLE
Associate Site model with Partner

### PR DESCRIPTION
**Pull request by mjfrey**
Associate Site model with Partner model to allow for multi-tenant support.

LEARNER-1117

### CODE REVIEW

File: course_discovery/apps/core/models.py [hunk 1]
```diff
@@ -78,6 +79,7 @@ class Partner(TimeStampedModel):
     oidc_key = models.CharField(max_length=255, null=True, verbose_name=_('OpenID Connect Key'))
     oidc_secret = models.CharField(max_length=255, null=True, verbose_name=_('OpenID Connect Secret'))
     studio_url = models.URLField(max_length=255, null=True, blank=True, verbose_name=_('Studio URL'))
+    site = models.OneToOneField(Site, null=True, blank=True)
```
Comment (vkaracic) [inline course_discovery/apps/core/models.py]: A partner can be without a site attached to it?

File: course_discovery/apps/publisher_comments/tests/factories.py [hunk 2]
```diff
@@ -1,16 +1,11 @@
 import factory
 from django.contrib.sites.models import Site
 
-from course_discovery.apps.core.tests.factories import UserFactory
+from course_discovery.apps.core.tests.factories import SiteFactory, UserFactory
 from course_discovery.apps.publisher.tests.factories import CourseRunFactory
 from course_discovery.apps.publisher_comments.models import Comments
 
 
-class SiteFactory(factory.DjangoModelFactory):  # pylint: disable=missing-docstring
```
Comment (vkaracic) [inline course_discovery/apps/publisher_comments/tests/factories.py:9]: Why the move to the upper file?

File: course_discovery/apps/core/models.py (hunk 1 above)
Comment (mjfrey) [reply course_discovery/apps/core/models.py]: Yes, for now, until we fix up the data.

File: course_discovery/apps/publisher_comments/tests/factories.py (hunk 2 above)
Comment (mjfrey) [reply course_discovery/apps/publisher_comments/tests/factories.py:9]: Should be in core and is not specific to publisher.

File: course_discovery/apps/core/models.py (hunk 1 above)
Comment (vkaracic) [reply course_discovery/apps/core/models.py]: Is there a ticket for that fix? I'd like to see what data needs to be fixed, cause I might be able to do it in the same migration file.

Comment (mjfrey) [reply course_discovery/apps/core/models.py]: We can just do this manually as there will only be a handful.  Can be done via the django admin.

File: course_discovery/apps/core/admin.py [hunk 3]
```diff
@@ -59,4 +59,4 @@ class PartnerAdmin(admin.ModelAdmin):
     )
     list_display = ('name', 'short_code',)
     ordering = ('name', 'short_code',)
-    search_fields = ('name', 'short_code',)
+    search_fields = ('name', 'short_code', 'site')
```
Comment (clintonb) [inline course_discovery/apps/core/admin.py]: This field should be in the list display. Given the handful of sites, there is no need for search.

File: course_discovery/apps/core/migrations/0004_partner_site.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    dependencies = [
+        ('sites', '0002_alter_domain_unique'),
+        ('core', '0003_auto_20170522_0856'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='partner',
+            name='site',
+            field=models.OneToOneField(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='sites.Site'),
```
Comment (clintonb) [inline course_discovery/apps/core/migrations/0004_partner_site.py]: Cascade is not good. We don't want to delete all of the data. We should disable this.

File: course_discovery/apps/core/tests/factories.py [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
 
+
 USER_PASSWORD = 'password'
 
 
+class SiteFactory(factory.DjangoModelFactory):
+    class Meta(object):
+        model = Site
+
+    domain = factory.fuzzy.FuzzyText(suffix='.fake')
```
Comment (clintonb) [inline course_discovery/apps/core/tests/factories.py]: Fuzzy is deprecated. Use Faker.

File: course_discovery/apps/core/tests/factories.py [hunk 6]
```diff
@@ -1,12 +1,21 @@
...
 
 from course_discovery.apps.core.models import Partner, User
 from course_discovery.apps.core.tests.utils import FuzzyUrlRoot
 
+
 USER_PASSWORD = 'password'
 
 
+class SiteFactory(factory.DjangoModelFactory):
+    class Meta(object):
```
Comment (clintonb) [inline course_discovery/apps/core/tests/factories.py]: No need for the parenthetical in Python 3.5.

**Review by clintonb — CHANGES_REQUESTED** [review]

File: course_discovery/apps/core/migrations/0004_partner_site.py (hunk 4 above)
Comment (clintonb) [reply course_discovery/apps/core/migrations/0004_partner_site.py]: https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.ForeignKey.on_delete

PROTECT is what we want here.

**Comment by vkaracic** [comment]
@clintonb comments addressed.

File: course_discovery/apps/core/tests/factories.py [hunk 7]
```diff
@@ -1,12 +1,21 @@
...
 
+
 USER_PASSWORD = 'password'
 
 
+class SiteFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = Site
+
+    domain = factory.fuzzy.FuzzyText(suffix='.fake')
```
Comment (clintonb) [inline course_discovery/apps/core/tests/factories.py]: Same comment as before: Fuzzy is deprecated. Use Faker. See https://github.com/edx/credentials/blob/master/credentials/apps/core/tests/factories.py#L25-L30.

File: course_discovery/apps/core/tests/factories.py [hunk 8]
```diff
@@ -31,11 +40,12 @@ class PartnerFactory(factory.DjangoModelFactory):
     programs_api_url = '{root}/api/programs/v1/'.format(root=FuzzyUrlRoot().fuzz())
     marketing_site_api_url = '{root}/api/courses/v1/'.format(root=FuzzyUrlRoot().fuzz())
     marketing_site_url_root = '{root}/'.format(root=FuzzyUrlRoot().fuzz())
-    marketing_site_api_username = FuzzyText().fuzz()
-    marketing_site_api_password = FuzzyText().fuzz()
+    marketing_site_api_username = factory.fuzzy.FuzzyText().fuzz()
```
Comment (clintonb) [inline course_discovery/apps/core/tests/factories.py]: Use Faker. See https://github.com/edx/credentials/blob/master/credentials/apps/core/tests/factories.py#L33 for examples.

**Review by clintonb — APPROVED** [review]
Use Faker.

**Pull request merged**